resources/session04/mashup.py

```diff
@@ -0,0 +1,201 @@
+    total = 0
+    high_score = 0
+    average = 0
+    for row in inspection_rows:
+        strval = clean_data(row.find_all('td')[2])
+        try:
+            intval = int(strval)
+        except (ValueError, TypeError):
+            samples -= 1
+        else:
+            total += intval
+            high_score = intval if intval > high_score else high_score
+
+    if samples:
+        average = total/float(samples)
+    data = {
+        u'Average Score': average,
+        u'High Score': high_score,
+        u'Total Inspections': samples
+    }
+    return data
+
+
+def result_generator(count):
+    use_params = {
+        'Inspection_Start': '2/1/2013',
+        'Inspection_End': '2/1/2015',
+        'Zip_Code': '98101'
+    }
+    # html = get_inspection_page(**use_params)
+    html = load_inspection_page('inspection_page.html')
+    parsed = parse_source(html)
+    content_col = parsed.find("td", id="contentcol")
+    data_list = restaurant_data_generator(content_col)
+    for data_div in data_list[:count]:
+        metadata = extract_restaurant_metadata(data_div)
+        inspection_data = get_score_data(data_div)
+        metadata.update(inspection_data)
+        yield metadata
+
+def sorted_results(key, count, reverse):
+    #translate key
+    if key == "highscore":
+        sort_by = "High Score"
+    elif key == "average":
+        sort_by = "Average Score"
+    elif key == "mostinpection":
+        sort_by = "Total Inspections"
+
+    use_params = {
+        'Inspection_Start': '2/1/2013',
+        'Inspection_End': '2/1/2015',
+        'Zip_Code': '98101'
+    }
+    html = load_inspection_page('inspection_page.html')
+    parsed = parse_source(html)
+    content_col = parsed.find("td", id="contentcol")
+    data_list = restaurant_data_generator(content_col)
+    metadata_list = []
+    for data_div in data_list:
+        metadata = extract_restaurant_metadata(data_div)
+        inspection_data = get_score_data(data_div)
+        metadata.update(inspection_data)
+        metadata_list.append(metadata)
+    #take into account of sort is sorted asc by default
+    rev = not reverse
+    sorted_list = sorted(metadata_list, key=itemgetter(sort_by), reverse=rev)
+    return sorted_list[0:count]
+
+
+
+def get_geojson(result):
+    address = " ".join(result.get('Address', ''))
+    if not address:
+        return None
+    geocoded = geocoder.google(address)
+    geojson = geocoded.geojson
+    inspection_data = {}
+    use_keys = (
+        'Business Name', 'Average Score', 'Total Inspections', 'High Score'
+    )
+    for key, val in result.items():
+        if key not in use_keys:
+            continue
+        if isinstance(val, list):
+            val = " ".join(val)
+        inspection_data[key] = val
+    geojson['properties'] = inspection_data
+    return geojson
+
+
+if __name__ == '__main__':
+    args = sys.argv
+    sort = args[1]
+    num_result = int(args[2])
+    reverse = True if len(args) == 4 else False
```

Comment:
This works, but is highly dependent on the ordering of arguments at the command line.  What happens if the value passed for `sort` is in the third position instead of the second?  What happens if there is no value for `num_result`, or if it comes out of order?  There are lots of ways that usage errors could cause trouble for you here, so you have a couple of options:

1.  Manually validate all inputs and provide useful error messages if the way the user uses your script doesn't match your expectations.  
2.  Use a command-line parsing tool from the standard library ([`argparse`](https://docs.python.org/3/howto/argparse.html#id1)) to do validation and error handling for you.
3.  Use a more advanced command-line application module (like [`click`](http://click.pocoo.org/6/)) to build a more fluent command line application.